File: Src/java/model/src/main/java/org/hl7/cql/model/ClassType.java
package org.hl7.cql.model;

import java.util.*;

public class ClassType extends DataType implements NamedType {

    public ClassType(String name, DataType baseType, Collection<ClassTypeElement> elements, Collection<TypeParameter> parameters) {
        super(baseType);

        if (name == null || name.equals("")) {
            throw new IllegalArgumentException("name is null");
        }

        this.name = name;

        if (parameters != null) {
            this.genericParameters.addAll(parameters);
        }

        if (elements != null) {
            this.elements.addAll(elements);
        }
    }

    public ClassType() {
        this(null, null, null, null);
    }

    public ClassType(String name) {
        this(name, null, null, null);
    }

    public ClassType(String name, DataType baseType) {
        this(name, baseType, null, null);
    }

    public ClassType(String name, DataType baseType, Collection<ClassTypeElement> elements) {
        this(name, baseType, elements, null);
    }

    private String name;
    public String getName() {
        return this.name;
    }

    public String getNamespace() {
        if (this.name != null) {
            int qualifierIndex = this.name.indexOf('.');//TODO Should this not be the last occurrence rather than the first occurrence?
            if (qualifierIndex > 0) {
                return this.name.substring(0, qualifierIndex);
            }
        }

        return "";
    }

    public String getSimpleName() {
        if (this.name != null) {
            int qualifierIndex = this.name.indexOf('.');//TODO Should this not be the last occurrence rather than the first occurrence?
            if (qualifierIndex > 0) {
                return this.name.substring(qualifierIndex + 1);
            }
        }

        return this.name;
    }

    private String identifier;
    public String getIdentifier() { return identifier; }
    public void setIdentifier(String identifier) { this.identifier = identifier; }

    private String label;
    public String getLabel() { return label; }
    public void setLabel(String label) { this.label = label; }

    private boolean retrievable;
    public boolean isRetrievable() {
        return retrievable;
    }
    public void setRetrievable(boolean retrievable) {
        this.retrievable = retrievable;
    }

    private String primaryCodePath;
    public String getPrimaryCodePath() { return primaryCodePath; }
    public void setPrimaryCodePath(String primaryCodePath) {
        this.primaryCodePath = primaryCodePath;
    }

    private String primaryValueSetPath;
    public String getPrimaryValueSetPath() { return primaryValueSetPath; }
    public void setPrimaryValueSetPath(String primaryValueSetPath) {
        this.primaryValueSetPath = primaryValueSetPath;
    }

    private List<Relationship> relationships = new ArrayList<>();
    public Iterable<Relationship> getRelationships() {
        return relationships;
    }

    public void addRelationship(Relationship relationship) {
        relationships.add(relationship);
    }

    private List<Relationship> targetRelationships = new ArrayList<>();
    public Iterable<Relationship> getTargetRelationships() {
        return targetRelationships;
    }

    public void addTargetRelationship(Relationship relationship) {
        targetRelationships.add(relationship);
    }

    /**
     * Generic class parameters such 'S', 'T extends MyType'.
     */
    private List<TypeParameter> genericParameters = new ArrayList<>();

    /**
     * Returns the generic parameters for the generic type. For instance,
     * for the generic type Map&lt;K,V extends Person&gt;, two generic parameters
     * will be returned: K and V extends Person. The latter parameter has a constraint
     * restricting the type of the bound type to be a valid subtype of Person.
     *
     * @return Class' generic parameters
     */
    public List<TypeParameter> getGenericParameters() {
        return genericParameters;
    }

    /**
     * Sets the generic parameters for the generic type. For instance,
     * for the generic type Map&lt;K,V extends Person&gt;, two generic parameters
     * should be set: K and V extends Person. The latter parameter has a constraint
     * restricting the type of the bound type to be a valid subtype of Person.
     *
     * @param genericParameters
     */
    public void setGenericParameters(List<TypeParameter> genericParameters) {
        this.genericParameters = genericParameters;
    }

    /**
     * Adds a parameter declaration to the generic type.
     *
     * @param genericParameter
     */
    public void addGenericParameter(TypeParameter genericParameter) {
        this.genericParameters.add(genericParameter);
    }

    /**
     * Adds collection of type parameters to existing set.
     * @param parameters
     */
    public void addGenericParameter(Collection<TypeParameter> parameters) {
        for (TypeParameter parameter : parameters) {
            internalAddParameter(parameter);
        }

        sortedElements = null;
        tupleType = null;
    }

    /**
     * Returns the parameter with the given parameter identifier.
     * If not found in the given class, it looks in the parent class.
     *
     * @param identifier
     * @return Generic parameter with the given name in the current class or in the base class. Null if none found.
     */
    public TypeParameter getGenericParameterByIdentifier(String identifier) {
        return getGenericParameterByIdentifier(identifier, false);
    }

    /**
     * Returns the parameter with the given parameter identifier.
     * If inCurrentClassOnly is false, if not found in the given class, then it looks in the parent class.
     * If inCurrentClassOnly is true, only looks for parameter in the given class.
     *
     * @param identifier
     * @param inCurrentClassOnly
     * @return Class' generic parameter
     */
    public TypeParameter getGenericParameterByIdentifier(String identifier, boolean inCurrentClassOnly) {
        TypeParameter param = null;
        for(TypeParameter genericParameter: genericParameters) {
            if(identifier.equalsIgnoreCase(genericParameter.getIdentifier())) {
                param = genericParameter;
                break;
            }
        }
        if(!inCurrentClassOnly && param == null) {
            if (param == null && getBaseType() instanceof ClassType) {
                param = ((ClassType) getBaseType()).getGenericParameterByIdentifier(identifier);
            }
        }
        return param;
    }

    private List<ClassTypeElement> elements = new ArrayList<ClassTypeElement>();
    private List<ClassTypeElement> sortedElements = null;
    private LinkedHashMap<String, ClassTypeElement> baseElementMap = null;

    public List<ClassTypeElement> getElements() {
        return elements;
    }

    private LinkedHashMap<String, ClassTypeElement> getBaseElementMap() {
        if (baseElementMap == null) {
            baseElementMap = new LinkedHashMap<>();
            if (getBaseType() instanceof ClassType) {
                ((ClassType)getBaseType()).gatherElements(baseElementMap);
            }
        }

        return baseElementMap;
    }

    private void gatherElements(LinkedHashMap<String, ClassTypeElement> elementMap) {
        if (getBaseType() instanceof ClassType) {
            ((ClassType)getBaseType()).gatherElements(elementMap);
        }

        for (ClassTypeElement element : elements) {
            elementMap.put(element.getName(), element);
        }
    }

    public List<ClassTypeElement> getAllElements() {
        // Get the baseClass elements into a map by name
        LinkedHashMap<String, ClassTypeElement> elementMap = new LinkedHashMap<>(getBaseElementMap());

        // Add this class's elements, overwriting baseClass definitions where applicable
        for (ClassTypeElement el : elements) {
            elementMap.put(el.getName(), el);
        }

        return new ArrayList<>(elementMap.values());
    }

    private void internalAddElement(ClassTypeElement element) {
        ClassTypeElement existingElement = getBaseElementMap().get(element.getName());
        if (
            existingElement != null && !(existingElement.getType() instanceof TypeParameter)
            && (
                !(
                    element.getType().isSubTypeOf(existingElement.getType())
                    || (
                        existingElement.getType() instanceof ListType
                        && element.getType().isSubTypeOf(((ListType)existingElement.getType()).getElementType())
                    )
                )
            )
        ) {
            throw new InvalidRedeclarationException(this, existingElement, element);
        }

        this.elements.add(element);
    }

    private void internalAddParameter(TypeParameter parameter) {
        //TODO Flesh out and retain method only if needed.

        this.genericParameters.add(parameter);
    }

    public void addElement(ClassTypeElement element)
    {
        internalAddElement(element);
        sortedElements = null;
        tupleType = null;
    }

    public void addElements(Collection<ClassTypeElement> elements) {
        for (ClassTypeElement element : elements) {
            internalAddElement(element);
        }

        sortedElements = null;
        tupleType = null;
    }

    private List<ClassTypeElement> getSortedElements() {
        if (sortedElements == null) {
            sortedElements = new ArrayList<>(elements);
            Collections.sort(sortedElements, (left, right) -> left.getName().compareTo(right.getName()));
        }

        return sortedElements;
    }

    @Override
    public int hashCode() {
        return this.name.hashCode();
    }

    @Override
    public boolean equals(Object o) {
        if (o instanceof ClassType) {
            ClassType that = (ClassType)o;
            return this.name.equals(that.name);
        }

        return false;
    }

    @Override
    public String toString() {
        return this.name;
    }

    @Override
    public String toLabel() {
        return this.label == null ? this.name : this.label;
    }

    private TupleType tupleType;
    public TupleType getTupleType() {
        if (tupleType == null) {
            tupleType = buildTupleType();
        }

        return tupleType;
    }

    private void addTupleElements(ClassType classType, LinkedHashMap<String, TupleTypeElement> elements) {
        // Add base elements first
        DataType baseType = classType.getBaseType();
        if (baseType instanceof ClassType) {
            addTupleElements((ClassType)baseType, elements);
        }

        for (ClassTypeElement element : classType.getElements()) {
            if (!element.isProhibited()) {
                TupleTypeElement tupleElement = new TupleTypeElement(element.getName(), element.getType());
                elements.put(tupleElement.getName(), tupleElement);
            }
        }
    }

    private TupleType buildTupleType() {
        LinkedHashMap<String, TupleTypeElement> tupleElements = new LinkedHashMap<>();

        addTupleElements(this, tupleElements);

        return new TupleType(tupleElements.values());
    }

    @Override
    public boolean isCompatibleWith(DataType other) {
        if (other instanceof TupleType) {
            TupleType tupleType = (TupleType)other;
            return getTupleType().equals(tupleType);
        // Github #115: It's incorrect for a class type to be considered compatible with another class type on the basis of the inferred tuple type alone.
        //} else if (other instanceof ClassType) {
        //    ClassType classType = (ClassType)other;
        //    return getTupleType().equals(classType.getTupleType());
        }

        return super.isCompatibleWith(other);
    }

    @Override
    public boolean isGeneric() {
        return genericParameters != null && genericParameters.size() > 0;
    }

    @Override
    public boolean isInstantiable(DataType callType, InstantiationContext context) {
        if (callType instanceof ClassType) {
            ClassType classType = (ClassType)callType;
            if (elements.size() == classType.elements.size()) {
                List<ClassTypeElement> theseElements = getSortedElements();
                List<ClassTypeElement> thoseElements = classType.getSortedElements();
                for (int i = 0; i < theseElements.size(); i++) {
                    if (!(theseElements.get(i).getName().equals(thoseElements.get(i).getName())
                            && theseElements.get(i).getType().isInstantiable(thoseElements.get(i).getType(), context))) {
                        return false;
                    }
                }

                return true;
            }
        }

        return false;
    }

    @Override
    public DataType instantiate(InstantiationContext context) {
        if (!isGeneric()) {
            return this;
        }

        ClassType result = new ClassType(getName(), getBaseType());
        for (int i = 0; i < elements.size(); i++) {
            result.addElement(new ClassTypeElement(elements.get(i).getName(), elements.get(i).getType().instantiate(context)));
        }

        return result;
    }
}

File: Src/java/model/src/main/java/org/hl7/cql/model/ClassType.kt
package org.hl7.cql.model

import java.util.*

@Suppress("TooManyFunctions", "ComplexCondition")
open class ClassType
@JvmOverloads
constructor(
    final override val name: String,
    baseType: DataType? = null,
    val elements: MutableList<ClassTypeElement> = mutableListOf(),
    /**
     * Generic class parameters such 'S', 'T extends MyType'.
     *
     * For instance, for the generic type Map&lt;K,V extends Person&gt;, two generic parameters will
     * be returned: K and V extends Person. The latter parameter has a constraint restricting the
     * type of the bound type to be a valid subtype of Person.
     */
    var genericParameters: MutableList<TypeParameter> = mutableListOf()
) : BaseDataType(baseType), NamedType {
    init {
        require(name.isNotEmpty()) { "A class type must have a name." }
    }

    override val namespace: String
        get() {
            // TODO Should this not be the last occurrence rather than the first occurrence?
            val qualifierIndex = name.indexOf('.')
            return if (qualifierIndex > 0) {
                name.substring(0, qualifierIndex)
            } else ""
        }

    override val simpleName: String
        get() {
            // TODO Should this not be the last occurrence rather than the first occurrence?
            val qualifierIndex = name.indexOf('.')
            return if (qualifierIndex > 0) {
                name.substring(qualifierIndex + 1)
            } else name
        }

    var identifier: String? = null
    var label: String? = null
    override var target: String? = null
    var isRetrievable: Boolean = false

    var primaryCodePath: String? = null

    var primaryValueSetPath: String? = null

    private val relationships: MutableList<Relationship> = ArrayList()

    fun getRelationships(): List<Relationship> {
        return relationships
    }

    fun addRelationship(relationship: Relationship) {
        relationships.add(relationship)
    }

    private val targetRelationships: MutableList<Relationship> = ArrayList()

    fun getTargetRelationships(): List<Relationship> {
        return targetRelationships
    }

    fun addTargetRelationship(relationship: Relationship) {
        targetRelationships.add(relationship)
    }

    private val searches: MutableList<SearchType> = ArrayList()

    fun getSearches(): List<SearchType> {
        return searches
    }

    fun addSearch(search: SearchType) {
        searches.add(search)
    }

    fun findSearch(searchPath: String): SearchType? {
        return searches.firstOrNull() { it.name == searchPath }
    }

    /**
     * Adds a parameter declaration to the generic type.
     *
     * @param genericParameter
     */
    fun addGenericParameter(genericParameter: TypeParameter) {
        genericParameters.add(genericParameter)
    }

    /**
     * Adds collection of type parameters to existing set.
     *
     * @param parameters
     */
    fun addGenericParameter(parameters: Collection<TypeParameter>) {
        for (parameter in parameters) {
            internalAddParameter(parameter)
        }
    }

    /**
     * Returns the parameter with the given parameter identifier. If not found in the given class,
     * it looks in the parent class.
     *
     * @param identifier
     * @return Generic parameter with the given name in the current class or in the base class. Null
     *   if none found.
     */
    fun getGenericParameterByIdentifier(identifier: String): TypeParameter? {
        return getGenericParameterByIdentifier(identifier, false)
    }

    /**
     * Returns the parameter with the given parameter identifier. If inCurrentClassOnly is false, if
     * not found in the given class, then it looks in the parent class. If inCurrentClassOnly is
     * true, only looks for parameter in the given class.
     *
     * @param identifier
     * @param inCurrentClassOnly
     * @return Class' generic parameter
     */
    fun getGenericParameterByIdentifier(
        identifier: String,
        inCurrentClassOnly: Boolean
    ): TypeParameter? {
        var param: TypeParameter? = null
        for (genericParameter in genericParameters) {
            if (identifier.equals(genericParameter.identifier, ignoreCase = true)) {
                param = genericParameter
                break
            }
        }
        if (!inCurrentClassOnly && param == null && baseType is ClassType) {
            param = (baseType as ClassType).getGenericParameterByIdentifier(identifier)
        }
        return param
    }

    val sortedElements: List<ClassTypeElement>
        get() = elements.sortedWith { o1, o2 -> o1.name.compareTo(o2.name) }

    private var baseElementMap: LinkedHashMap<String, ClassTypeElement>? = null
        get() {
            if (field == null) {
                field = LinkedHashMap()
                if (baseType is ClassType) {
                    (baseType as ClassType).gatherElements(field!!)
                }
            }

            return field
        }

    private fun gatherElements(elementMap: LinkedHashMap<String, ClassTypeElement>) {
        if (baseType is ClassType) {
            (baseType as ClassType).gatherElements(elementMap)
        }

        for (element in elements) {
            elementMap[element.name] = element
        }
    }

    val allElements: List<ClassTypeElement>
        get() {
            // Get the baseClass elements into a map by name
            val elementMap = LinkedHashMap(baseElementMap)

            // Add this class's elements, overwriting baseClass definitions where applicable
            for (el in elements) {
                elementMap[el.name] = el
            }

            return elementMap.values.toList()
        }

    private fun internalAddElement(element: ClassTypeElement) {
        val existingElement = baseElementMap!![element.name]
        if (
            existingElement != null &&
                (existingElement.type !is TypeParameter) &&
                (!(element.type.isSubTypeOf(existingElement.type) ||
                    (existingElement.type is ListType &&
                        element.type.isSubTypeOf(existingElement.type.elementType)) ||
                    (existingElement.type is IntervalType &&
                        element.type.isSubTypeOf(existingElement.type.pointType)) ||
                    (existingElement.type is ChoiceType &&
                        element.type.isCompatibleWith(existingElement.type))))
        ) {
            throw InvalidRedeclarationException(this, existingElement, element)
        }

        elements.add(element)
    }

    private fun internalAddParameter(parameter: TypeParameter) {
        // TODO Flesh out and retain method only if needed.

        genericParameters.add(parameter)
    }

    fun addElement(element: ClassTypeElement) {
        internalAddElement(element)
    }

    fun addElements(elements: Collection<ClassTypeElement>) {
        for (element in elements) {
            internalAddElement(element)
        }
    }

    override fun hashCode(): Int {
        return name.hashCode()
    }

    override fun equals(other: Any?): Boolean {
        if (other is ClassType) {
            return this.name == other.name
        }

        return false
    }

    override fun toString(): String {
        return this.name
    }

    override fun toLabel(): String {
        return if (this.label == null) this.name else label!!
    }

    val tupleType: TupleType
        get() = buildTupleType()

    private fun addTupleElements(
        classType: ClassType,
        elements: LinkedHashMap<String, TupleTypeElement>
    ) {
        // Add base elements first
        val baseType = classType.baseType
        if (baseType is ClassType) {
            addTupleElements(baseType, elements)
        }

        for ((name1, type, prohibited) in classType.elements) {
            if (!prohibited) {
                val tupleElement = TupleTypeElement(name1, type, false)
                elements[tupleElement.name] = tupleElement
            }
        }
    }

    private fun buildTupleType(): TupleType {
        val tupleElements = LinkedHashMap<String, TupleTypeElement>()

        addTupleElements(this, tupleElements)

        return TupleType(tupleElements.values)
    }

    override fun isCompatibleWith(other: DataType): Boolean {
        if (other is TupleType) {
            return tupleType == other
            // GitHub #115: It's incorrect for a class type to be considered compatible with another
            // class type on the
            // basis of the inferred tuple type alone.
            // } else if (other instanceof ClassType) {
            //    ClassType classType = (ClassType)other;
            //    return getTupleType().equals(classType.getTupleType());
        }

        return super.isCompatibleWith(other)
    }

    override val isGeneric: Boolean
        get() = genericParameters.isNotEmpty()

    override fun isInstantiable(callType: DataType, context: InstantiationContext): Boolean {
        return if (callType is ClassType && callType.elements.size == elements.size) {
            sortedElements.zip(callType.sortedElements).all { (thisElement, thatElement) ->
                thisElement.type.isInstantiable(thatElement.type, context)
            }
        } else false
    }

    override fun instantiate(context: InstantiationContext): DataType {
        if (!isGeneric) {
            return this
        }

        return ClassType(
            name,
            baseType,
            elements.map { ClassTypeElement(it.name, it.type.instantiate(context)) }.toMutableList()
        )
    }
}